File: src/main/java/edu/practice/datastructures/linkedlist/SinglyLinkedList.java
package edu.practice.datastructures.linkedlist;

import java.util.NoSuchElementException;
import java.util.Objects;

public class SinglyLinkedList<E> {

    private static class Node<E> {
        E data;
        Node<E> next;

        Node(E data, Node<E> next) {
            this.data = data;
            this.next = next;
        }
    }

    private int size;
    private Node<E> first;
    private Node<E> last;

    public SinglyLinkedList() {// Initialize empty Linked List
    }

    public int size() { return size; }

    public boolean isEmpty() {
        return (size == 0);
    }

    public void addFirst(E item) {
        first = new Node<>(item, first);

        if (last == null) {
            last = first;
        }

        size++;
    }

    public void addLast(E item) {
        final Node<E> newNode = new Node<>(item, null);

        if (last == null) {
            last = newNode;
            first = newNode;
        } else {
            last.next = newNode;
            last = last.next;
        }

        size++;
    }

    public E getFirst() {
        if (first == null) {
            throw new NoSuchElementException();
        }

        return first.data;
    }

    public E getLast() {
        if (last == null) {
            throw new NoSuchElementException();
        }

        return last.data;
    }

    public E removeFirst() {
        final E item = first.data;
        first = first.next;

        // First is now pointing to previously 2nd element of list, so we check if there was no other element
        if (first == null) {
            last = null;
        }

        size--;

        return item;
    }

    public E removeLast() {
        final E item = last.data;

        if (first != last) { // More than one element in the list
            Node<E> currentNode = first;
            while (currentNode.next != last)
                currentNode = currentNode.next;

            currentNode.next = null;
            last = currentNode;
        } else {
            first = null;
            last = null;
        }

        size--;

        return item;
    }

    private boolean existsInList(E item, Node<E> firstNode) {
        if (firstNode == null) {
            return false;
        }

        if (Objects.equals(item, firstNode.data)) {
            return true;
        }

        return existsInList(item, firstNode.next);
    }

    public boolean contains(E item) {
        return existsInList(item, first);
    }
}

File: src/main/kotlin/datastructures/linkedlist/SinglyLinkedList.kt
package edu.practice.datastructures.linkedlist

open class SinglyLinkedList<E>: LinkedList<E> {

    private data class Node<E>(val data: E, var next: Node<E>? = null)

    private var size = 0

    private var first: Node<E>? = null
    private var last: Node<E>? = null

    override fun size(): Int = size

    override fun isEmpty(): Boolean = (size == 0)

    override fun getFirst(): E {
        return first?.data ?: throw NoSuchElementException("Linked List is Empty")
    }

    override fun getLast(): E {
        return last?.data ?: throw NoSuchElementException("Linked List is Empty")
    }

    override fun addFirst(item: E) {
        val newNode = Node(item, first)
        first = newNode

        if (last == null) {
            last = newNode
        }

        size++
    }

    override fun addLast(item: E) {
        val newNode = Node(item)

        if (last == null) {
            first = newNode
            last = newNode
        } else {
            last?.next = newNode
            last = newNode
        }

        size++
    }

    override fun removeFirst(): E {
        val item = getFirst()

        first = first?.next
        if (first == null) {
            last = null
        }

        size--
        return item
    }

    override fun removeLast(): E {
        val item = getLast()

        if (first != last) {
            // > 1 element in the list
            var currentNode = first
            while (currentNode?.next != last) {
                currentNode = currentNode?.next
            }

            currentNode?.next = null
            last = currentNode
        } else {
            first = null
            last = null
        }

        size--
        return item
    }

    private fun existsInList(item: E, firstNode: Node<E>? = first): Boolean {
        return firstNode?.let {
            it.data == item || existsInList(item, it.next)
        } ?: false
    }

    override fun contains(item: E): Boolean {
        return existsInList(item)
    }
}